internal/internal_workflow_client.go

```diff
@@ -734,6 +735,55 @@ func (wc *WorkflowClient) QueryWorkflow(ctx context.Context, workflowID string,
 	})
 }
 
+// UpdateWorkflowWithOptionsRequest is the request to UpdateWorkflowWithOptions
+type UpdateWorkflowWithOptionsRequest struct {
+	// UpdateID is an application-layer identifier for the requested update. It
+	// must be unique within the scope of a Namespace+WorkflowID+RunID.
+	UpdateID string
+
+	// WorkflowID is a required field indicating the workflow which should be
+	// updated.
+	WorkflowID string
+
+	// RunID is an optional field used to identify a specific run of the target
+	// workflow.  If RunID is not provided the latest run will be used.
+	RunID string
+
+	// UpdateName is a required field which specifies the update you want to run.
+	// See comments at workflow.SetUpdateHandler(ctx Context, updateName string, handler interface{}, opts UpdateHandlerOptions)
+	// for more details on how to setup update handlers within the target workflow.
+	UpdateName string
+
+	// Args is an optional field used to identify the arguments passed to the
+	// update.
+	Args []interface{}
+
+	// FirstExecutionRunID specifies the RunID expected to identify the first
+	// run in the workflow execution chain. If this expectation does not match
+	// then the server will reject the update request with an error.
+	FirstExecutionRunID string
+
+	// this isn't upstream in API yet
+	// WaitFor enumspb.WorkflowExecutionUpdateWaitEvent
+}
+
+// WorkflowUpdateHandle is a handle to a workflow execution update
+// process. The update may or may not have completed so an instance of this type
+// functions simlarly to a Future with respect to the result value of the
+// update. If the update is rejected or returns an error, the Get function on
+// this type will return that error.
+type WorkflowUpdateHandle interface {
+	// UpdateRef() *updatepb.UpdateRef //once API lands
+	Get(ctx context.Context, valuePtrs ...interface{}) error
+}
+
+// updateHandle is a dumb implementation of WorkflowExecutionUpdateHandle that
+// only supports the case where the output value is aready known at construction
+// time.
+type updateHandle struct {
+	value interface{} // converter.EncodedValue | error
```

Comment:
nit: I personally prefer a strongly typed version of this that saves the type switch but I don't have a strong opinion.

Reply:
I agree, I think two mutually exclusive fields for success and failure here is much more reasonable.

Reply:
Switched to a product type